src/lru/mod.rs

```diff
@@ -57,14 +53,234 @@ where
     // to it (which means that your pin_count will be 0). If a Page/FrameID does not
     // have any references we can remove from cache.
     pub fn unpin(&mut self, id: &T) {
-        if !self.elements.contains(id) {
-            self.elements.insert(0, id.clone());
-        }
+        self.list.push_front(id.clone(), ());
     }
 
     /// Returns the number of frames that are currently in the LRUReplacer.
     pub fn size(&self) -> usize {
-        self.elements.len()
+        self.list.size()
+    }
+
+    /// Returns the elements in the LRUReplacer.
+    pub fn elements(&self) -> Vec<T> {
+        self.list
+            .clone()
+            .into_iter()
+            .map(|node| node.key.clone())
+            .collect()
+    }
+}
+
+use std::iter::IntoIterator;
+use std::{collections::HashMap, hash::Hash, ops::Deref, ops::DerefMut, ptr::NonNull};
+
+struct Node<K, V> {
+    key: K,
+    value: V,
+    prev: Option<NodePtr<K, V>>,
+    next: Option<NodePtr<K, V>>,
+}
+
+impl<K, V> Node<K, V> {
+    fn new(key: K, value: V) -> Self {
+        Self {
+            key,
+            value,
+            prev: None,
+            next: None,
+        }
+    }
+}
+
+struct NodePtr<K, V> {
+    ptr: NonNull<Node<K, V>>,
+}
+
+impl<K, V> NodePtr<K, V> {
+    fn new(node: &Node<K, V>) -> Self {
+        Self {
+            ptr: NonNull::from(node),
+        }
+    }
+}
+
+impl<K, V> Deref for NodePtr<K, V> {
+    type Target = Node<K, V>;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { self.ptr.as_ref() }
+    }
+}
+
+impl<K, V> DerefMut for NodePtr<K, V> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        unsafe { self.ptr.as_mut() }
+    }
+}
+
+impl<K, V> Clone for NodePtr<K, V> {
+    fn clone(&self) -> Self {
+        Self { ptr: self.ptr }
+    }
+}
+
+impl<K, V> Copy for NodePtr<K, V> {}
+
+impl<K: Hash, V> Hash for NodePtr<K, V> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        self.key.hash(state);
+    }
+}
+
+impl<K: PartialEq, V> PartialEq for NodePtr<K, V> {
+    fn eq(&self, other: &Self) -> bool {
+        self.key == other.key
+    }
+}
+
+impl<K: Eq, V> Eq for NodePtr<K, V> {}
+
+unsafe impl<K, V> Send for NodePtr<K, V> {}
+
+unsafe impl<K, V> Sync for NodePtr<K, V> {}
+
+#[derive(Clone)]
+struct LinkedHashList<K, V> {
+    map: HashMap<NodePtr<K, V>, NodePtr<K, V>>,
```

Comment:
This is a fatal mistake.:(
After implementation, I didn't notice that key required to implement the Clone trait. In my original assumption, key only needed to implement the Borrow trait, which would have a wider range of applications.
The role of map is to implement O (1) lookup elements, and realize key sharing by storing nodes pointing to nodes instead of the key itself.